pkg/resource/training_job/custom_set_update_input.go

```diff
@@ -0,0 +1,125 @@
+// 2. Customer only updates Profiler Config: Set the profiler rule configuration to nil to avoid validation error.
+// 3. Customer only updates Rule Configurations: Recreate the input for profiler Rule and set Profiler config to nil.
+//	  safer to do this because the "only add" behavior might reappear.
+
+func customSetUpdateInput(desired *resource, latest *resource, delta *ackcompare.Delta, input *svcsdk.UpdateTrainingJobInput) error {
+	if delta.DifferentAt("Spec.ProfilerConfig") && delta.DifferentAt("Spec.ProfilerRuleConfigurations") {
+		err := handleProfilerRuleConfig(desired, latest, input)
+		return err
+	}
+	if delta.DifferentAt("Spec.ProfilerConfig") && !delta.DifferentAt("Spec.ProfilerRuleConfigurations") {
+		input.SetProfilerRuleConfigurations(nil)
+		return nil
+	}
+	if delta.DifferentAt("Spec.ProfilerRuleConfigurations") && !delta.DifferentAt("Spec.ProfilerConfig") {
+		err := handleProfilerRuleConfig(desired, latest, input)
+		input.SetProfilerConfig(nil) // SM still assumes the profiler config is the same.
+		return err
+	}
+	return nil
+}
+
+// Update training job is post operation wrt to the profiler parameters.
+// Because of this only NEW rules can be specified.
+// In this function we check to see if any new profiler configurstions have been added.
+func handleProfilerRuleConfig(desired *resource, latest *resource, input *svcsdk.UpdateTrainingJobInput) error {
+	profilerRuleDesired := desired.ko.Spec.ProfilerRuleConfigurations
+	profilerRuleLatest := latest.ko.Spec.ProfilerRuleConfigurations
+
+	if ackcompare.IsNil(profilerRuleDesired) {
+		return errors.New("[ACK_SM] Cannot remove a profiler rule.")
+	}
+	if ackcompare.IsNil(profilerRuleLatest) {
+		return nil
+	}
+	if len(profilerRuleDesired) < len(profilerRuleLatest) {
+		return errors.New("[ACK_SM] Cannot remove a profiler rule.")
+	}
+
+	ruleMap := map[string]int{}
+	profilerRuleInput := []*svcsdk.ProfilerRuleConfiguration{}
+	for _, rule := range profilerRuleLatest {
+		if ackcompare.IsNotNil(rule) && ackcompare.IsNotNil(rule.RuleConfigurationName) {
+			ruleMap[*rule.RuleConfigurationName] = 1
+		}
+	}
+	for _, rule := range profilerRuleDesired {
+		if ackcompare.IsNotNil(rule) && ackcompare.IsNotNil(rule.RuleConfigurationName) {
+			_, present := ruleMap[*rule.RuleConfigurationName]
+			if !present {
+				profilerRuleInput = append(profilerRuleInput, convertProfileRuleType(rule))
+			}
+		}
+	}
+	input.SetProfilerRuleConfigurations(profilerRuleInput)
+	return nil
+}
+
+// Recreates input and sets disable profiler to true
+func handleProfilerRemoval(input *svcsdk.UpdateTrainingJobInput) {
+	input.SetProfilerRuleConfigurations(nil)
+	profilerConfig := svcsdk.ProfilerConfigForUpdate{}
+	profilerConfig.SetDisableProfiler(true)
+	input.SetProfilerConfig(&profilerConfig)
+}
+
+// Sagemaker and kubernetes types are not the same so the input has to be reconstructed.
+func convertProfileRuleType(rule *smv1alpha.ProfilerRuleConfiguration) *svcsdk.ProfilerRuleConfiguration {
```

Comment:
Is there a way to implicitly/explicity convert the Kubernetes type to AWS API type. Dont want to have this function as it is not maintainable.

Reply:
Unfortunately I dont think so